osu.Game/Beatmaps/Beatmap.cs

```diff
@@ -48,6 +48,44 @@ public class Beatmap<T> : IBeatmap<T>
 
         public virtual IEnumerable<BeatmapStatistic> GetStatistics() => Enumerable.Empty<BeatmapStatistic>();
 
+        public double GetMostCommonBeatLength()
+        {
+            // The last playable time in the beatmap - the last timing point extends to this time.
+            // Note: This is more accurate and may present different results because osu-stable didn't have the ability to calculate slider durations in this context.
+            double lastTime = HitObjects.LastOrDefault()?.GetEndTime() ?? ControlPointInfo.TimingPoints.LastOrDefault()?.Time ?? 0;
+
+            var beatLengthsAndDurations =
+                // Construct a set of (beatLength, duration) tuples for each individual timing point.
+                ControlPointInfo.TimingPoints.Select((t, i) =>
+                                {
+                                    if (t.Time > lastTime)
+                                        return (beatLength: t.BeatLength, 0);
+
+                                    var nextTime = i == ControlPointInfo.TimingPoints.Count - 1 ? lastTime : ControlPointInfo.TimingPoints[i + 1].Time;
+                                    return (beatLength: t.BeatLength, duration: nextTime - t.Time);
+                                })
+                                // Aggregate durations into a set of (beatLength, duration) tuples for each beat length
+                                .GroupBy(t => t.beatLength)
```

Comment:
Are there no floating point related issues when you group by `beatLength`? When checking two floating point numbers for equality (like in this case) it's usually best to introduce some epsilon.

Reply:
Decent point but not overly sure it's worth addressing. Fixing this is a little more tricky than it might seem I think (you could aggregate beat length in bins, but that will have weird effects on BPM due to taking the inverse).

Reply:
Could also introduce a `key` in the tuple that has the value `(int) (t.BeatLength * 1000)` and `.GroupBy(t => t.key)` if we only want 3 decimal-digit-precision (which should be enough?) of the beat length.

Reply:
That's essentially what I was proposing.